Framework/PythonInterface/plugins/algorithms/SetSampleFromLogs.py

```diff
@@ -0,0 +1,188 @@
+
+    def PyInit(self):
+        self.copyProperties("SetSample", PROPS_FOR_SETSAMPLE)
+        self.declareProperty("FindGeometry", True,
+                             "Whether to look for the 'Height' parameter in the logs")
+        self.declareProperty("FindSample", True,
+                             "Whether to look for the sample material in the logs")
+        self.declareProperty("FindEnvironment", True,
+                             "Whether to look for the sample container in the logs")
+
+    def _createMaterial(self, runObject):
+        '''Create the sample material by combining the supplied material information with things found in the logs'''
+        # grab the starting information from the algorithm property
+        material = _toDict(self.getProperty("Material").value)
+        # only look in the logs if the user asks for it
+        if self.getProperty("FindSample").value:  # SetSample calls it the material
+            if (not _hasValue(material, 'ChemicalFormula')) and _hasValue(runObject, "SampleFormula"):
+                self.log().information("Looking for 'SampleFormula' in logs")
+                material['ChemicalFormula'] = _getLogValue(runObject, 'SampleFormula').strip()
+            if (not _hasValue(material, "SampleMassDensity")) and _hasValue(runObject, "SampleDensity"):
+                self.log().information("Looking for 'SampleDensity', 'SampleMass', and 'SampleMassDensity' in logs")
+                value = _getLogValue(runObject, 'SampleDensity')
+                if value == 1.0 or value == 0.0:
+                    material['Mass'] = _getLogValue(runObject, 'SampleMass')
+                elif _hasValue(runObject, "SampleMass"):
+                    material['SampleMassDensity'] = value
+
+        # log the results and return
+        self.log().information('MATERIAL: ' + str(material))
+        return material
+
+    def _createGeometry(self, runObject, instrEnum):
+        # get height of sample from the logs
+        # this assumes the shape has a "Height" property
+        geometry = _toDict(self.getProperty("Geometry").value)
+
+        if self.getProperty("FindGeometry").value:
+            if not _hasValue(geometry, "Height"):
+                heightInContainerNames = ['HeightInContainer']
+                heightInContainerUnitsNames = ['HeightInContainerUnits']
+
+                # determine "special" logs for SNS instruments
+                if instrEnum.facility().name() == 'SNS':
+                    beamline = ''
+                    # TODO this would benefit from the beamline exposed through python
+                    if instrEnum.name() == 'NOMAD':
+                        beamline = 'BL1B'
+                    elif instrEnum.name() == 'POWGEN':
+                        beamline = 'BL11A'
+                    else:
+                        warningMsg = 'Do not know how to create lognames for "{}"'.format(instrEnum.name())
+                        self.log().warn(warningMsg)
```

Comment:
Just a comment here, once Python 3.10 becomes the standard, we should be able to `match-case` here to make it more concise